# Selecting samples for Synthetic data generation

## Importing libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
path = "/Users/saideepbunny/Projects/Application_Ranking_System"

## Reading the data

In [3]:
df = pd.read_csv(f"{path}/data/kaggle_linkedin_jd_skills_dataset/merged_selected_jd.csv")
df

job_link  \
0       https://www.linkedin.com/jobs/view/independent...   
1       https://www.linkedin.com/jobs/view/chef-de-cui...   
2       https://www.linkedin.com/jobs/view/cheese-spec...   
3       https://www.linkedin.com/jobs/view/retail-dist...   
4       https://www.linkedin.com/jobs/view/travel-rn-l...   
...                                                   ...   
125375  https://uk.linkedin.com/jobs/view/production-p...   
125376  https://www.linkedin.com/jobs/view/store-manag...   
125377  https://www.linkedin.com/jobs/view/sales-repre...   
125378  https://www.linkedin.com/jobs/view/ft-center-s...   
125379  https://www.linkedin.com/jobs/view/associate-a...   

                  last_processed_time got_summary got_ner is_being_worked  \
0       2024-01-21 07:40:00.308133+00           t       t               f   
1       2024-01-21 07:12:35.170823+00           t       t               f   
2       2024-01-21 06:01:14.692423+00           t       t               f   
3       2024-01-21 07:12:35.189652+00           t       t               f   
4       2024-01-21 07:12:37.537623+00           t       t               f   
...                               ...         ...     ...             ...   
125375  2024-01-20 04:01:08.913697+00           t       t               f   
125376  2024-01-20 05:12:47.255315+00           t       t               f   
125377  2024-01-20 13:46:06.129745+00           t       t               f   
125378  2024-01-20 13:46:10.289403+00           t       t               f   
125379   2024-01-20 15:20:54.32033+00           t       t               f   

                                 job_title                          company  \
0            Independent Real Estate Agent       Howard Hanna | Rand Realty   
1                          Chef de Cuisine               Goodwin Recruiting   
2                        Cheese Specialist                          Safeway   
3       RETAIL DISTRICT MANAGER UNASSIGNED                   Dollar General   
4               Travel RN - Long Term Care                CoreMedical Group   
...                                    ...                              ...   
125375                  Production Planner  Eat Real + PROPER | WARP Snacks   
125376           Store Manager in Training              O'Reilly Auto Parts   
125377                Sales Representative  American Fidelity Sales Careers   
125378         FT Center Store Manager (H)                        Food Lion   
125379           Associate Account Manager                          Lockton   

                             job_location  first_seen    search_city  \
0                    Englewood Cliffs, NJ  2024-01-16      Pinehurst   
1                          Brookfield, WI  2024-01-14      Milwaukee   
2                          Mount Airy, MD  2024-01-14      Frederick   
3                             Newport, NH  2024-01-14  New Hampshire   
4                          Barnstable, MA  2024-01-14     Barnstable   
...                                   ...         ...            ...   
125375  Nuneaton, England, United Kingdom  2024-01-14       Hastings   
125376                        Concord, NH  2024-01-14        Laconia   
125377                        Altoona, PA  2024-01-14   Pennsylvania   
125378                       Powhatan, VA  2024-01-14       Abingdon   
125379                    Minneapolis, MN  2024-01-15    Lake Forest   

        search_country             search_position   job_level job_type  \
0        United States           Real-Estate Clerk  Mid senior   Onsite   
1        United States                   Sous Chef  Mid senior   Onsite   
2        United States  Special Effects Specialist  Mid senior   Onsite   
3        United States         Pressure Supervisor  Mid senior   Onsite   
4        United States                  Nurse Head  Mid senior   Onsite   
...                ...                         ...         ...      ...   
125375  United Kingdom                   Scheduler  

In [4]:
df.isna().sum()

job_link                 0
last_processed_time      0
got_summary              0
got_ner                  0
is_being_worked          0
job_title                0
company                  3
job_location             2
first_seen               0
search_city              0
search_country           0
search_position          0
job_level                0
job_type                 0
job_skills             216
job_summary              0
id                       0
summary_len              0
dtype: int64

In [5]:
#drop those records with null values in job_skills column
df = df.dropna(subset=['job_skills']).copy()

## Examining job summaries

### All job summaries

In [6]:
print(df.loc[0]['job_summary'])

Who We Are
Rand Realty is a family-owned brokerage born in 1984 when our founder, Marsha opened her new real estate office in a small town in Hudson Valley. Her vision was to build a brokerage strongly devoted to its agents and deeply committed to the community combining expertise and eye for detail with heart, soul and kindness. More than 35 years later, we've grown to more than 1,000 premier local agents with offices throughout Northern New Jersey, Westchester, Connecticut, The Bronx, Rockland, Dutchess, Putnam and Orange counties. Despite our growth or really, by virtue of it and the great people who have joined and represent our brand, the characteristics that defined us then, continue to dictate and delineate who we are.
Today, Howard Hanna | Rand Realty is the preeminent real estate brokerage in New York, New Jersey and Connecticut, widely recognized as an industry leader for its agent training, career planning, technological innovation and market expertise. Build your career wit

In [7]:
print(df.loc[1]['job_summary'])

To Apply for this Job Click Here
Midwest Hospitality Group seeks a Chef de Cuisine to lead its Culinary Team. The ideal candidate will possess experience as a Chef de Cuisine and/or Executive Sous Chef in Independent Restaurants or Hotel Food & Beverage operations. Are you looking for a new challenge with a stable Company whose Culture is focused on providing the Guest with the highest quality Food and Service and offers growth opportunities? If so, please submit your resume today!
Benefits
Salary - $80,000 to $90,000
Health insurance
Life insurance
Free Parking
Shift Meals
PTO immediately
Company Discounts
Responsibilities
Ensure proper team coverage to execute a high level of production while keeping payroll costs in line
Oversee ordering, receiving, storage (including temperature-setting) usage, and rotation of food
Works with Maintenance in the repair of equipment and preventative maintenance programs
Assists with menu development and engineering
Leads culinary staff in the proper 

We can see from the above sample that few of the job summaries specifically have "Responsibilities" in the text. Since it is not possible to evaluate all the ~125k records to see if the job summary includes Responsibilities, we can simply filter the records based on the existence of the string "Responsibilities".

### Records containing "Responsibilities" in job summary

In [8]:
resp_df = df[df['job_summary'].str.contains("Responsibilities")]
resp_df

job_link  \
1       https://www.linkedin.com/jobs/view/chef-de-cui...   
7       https://www.linkedin.com/jobs/view/rad-tech-at...   
8       https://www.linkedin.com/jobs/view/retail-stoc...   
9       https://www.linkedin.com/jobs/view/grooming-sa...   
10      https://ca.linkedin.com/jobs/view/senior-produ...   
...                                                   ...   
125374  https://www.linkedin.com/jobs/view/quality-aud...   
125375  https://uk.linkedin.com/jobs/view/production-p...   
125377  https://www.linkedin.com/jobs/view/sales-repre...   
125378  https://www.linkedin.com/jobs/view/ft-center-s...   
125379  https://www.linkedin.com/jobs/view/associate-a...   

                  last_processed_time got_summary got_ner is_being_worked  \
1       2024-01-21 07:12:35.170823+00           t       t               f   
7       2024-01-19 14:39:33.898387+00           t       t               f   
8       2024-01-21 07:43:49.202444+00           t       t               f   
9       2024-01-21 07:43:50.084041+00           t       t               f   
10      2024-01-21 07:43:52.439593+00           t       t               f   
...                               ...         ...     ...             ...   
125374  2024-01-20 04:01:06.622679+00           t       t               f   
125375  2024-01-20 04:01:08.913697+00           t       t               f   
125377  2024-01-20 13:46:06.129745+00           t       t               f   
125378  2024-01-20 13:46:10.289403+00           t       t               f   
125379   2024-01-20 15:20:54.32033+00           t       t               f   

                          job_title                           company  \
1                   Chef de Cuisine                Goodwin Recruiting   
7                          Rad Tech  Virginia Mason Franciscan Health   
8         Retail Stocking Team Lead           Burlington Stores, Inc.   
9            Grooming Salon Manager                             Petco   
10           Senior Product Manager                       Robert Half   
...                             ...                               ...   
125374              Quality Auditor                           ANGARAI   
125375           Production Planner   Eat Real + PROPER | WARP Snacks   
125377         Sales Representative   American Fidelity Sales Careers   
125378  FT Center Store Manager (H)                         Food Lion   
125379    Associate Account Manager                           Lockton   

                             job_location  first_seen   search_city  \
1                          Brookfield, WI  2024-01-14     Milwaukee   
7                             Seattle, WA  2024-01-13        Summit   
8                            Chandler, AZ  2024-01-16      Chandler   
9                            Sarasota, FL  2024-01-16      Sarasota   
10               Calgary, Alberta, Canada  2024-01-14       Calgary   
...                                   ...         ...           ...   
125374                     Washington, DC  2024-01-14    Washington   
125375  Nuneaton, England, United Kingdom  2024-01-14      Hastings   
125377                        Altoona, PA  2024-01-14  Pennsylvania   
125378                       Powhatan, VA  2024-01-14      Abingdon   
125379                    Minneapolis, MN  2024-01-15   Lake Forest   

        search_country             search_position   job_level job_type  \
1        United States                   Sous Chef  Mid senior   Onsite   
7        United States                 Radiologist  Mid senior   Onsite   
8        United States     Pattern-Shop Supervisor  Mid senior   Onsite   
9        United States   Quality Assurance Monitor  Mid senior   Onsite   
10              Canada                        Tier  Mid senior   Onsite   
...                ...                         ...         ...      ...   
125374   United States          Standards Engineer  Mid senior   Onsite   
125375  United Kingdom                   Scheduler  Mid senio

In [9]:
for i in range(10):
    print(f"Record {i+1}:")
    print(resp_df.iloc[i]['job_summary'])
    print("\n================================================================================\n")

Record 1:
To Apply for this Job Click Here
Midwest Hospitality Group seeks a Chef de Cuisine to lead its Culinary Team. The ideal candidate will possess experience as a Chef de Cuisine and/or Executive Sous Chef in Independent Restaurants or Hotel Food & Beverage operations. Are you looking for a new challenge with a stable Company whose Culture is focused on providing the Guest with the highest quality Food and Service and offers growth opportunities? If so, please submit your resume today!
Benefits
Salary - $80,000 to $90,000
Health insurance
Life insurance
Free Parking
Shift Meals
PTO immediately
Company Discounts
Responsibilities
Ensure proper team coverage to execute a high level of production while keeping payroll costs in line
Oversee ordering, receiving, storage (including temperature-setting) usage, and rotation of food
Works with Maintenance in the repair of equipment and preventative maintenance programs
Assists with menu development and engineering
Leads culinary staff in t

## Statistics for summary length and skill count

In [10]:
#Splitting job skills by comma to get skill count
resp_df.loc[:,'skill_count'] = resp_df['job_skills'].apply(lambda x: len(x.split(","))).copy()


/var/folders/t5/tflj5hjs197_vv37fw2sdhdr0000gn/T/ipykernel_57662/3804758628.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  resp_df.loc[:,'skill_count'] = resp_df['job_skills'].apply(lambda x: len(x.split(","))).copy()


In [11]:
#getting max summary_len for each job title using groupby
resp_df[['job_title', 'job_summary', 'summary_len', 'skill_count']].groupby(['job_title', 'job_summary']).aggregate({"summary_len":"max", "skill_count":"max"}).reset_index()

job_title  \
0      2024 - Culinary Professionals - Assistant Kitc...   
1      2024 - Culinary Professionals - Dining Hall Se...   
2                     AE - Sales Leader (Full-Time) - US   
3                     AE - Sales Leader (Full-Time) - US   
4                     AE - Sales Leader (Full-Time) - US   
...                                                  ...   
37195                                    Yoga Instructor   
37196                                    Yoga Instructor   
37197                                    Yoga Instructor   
37198                                    Yoga Instructor   
37199                                    Yoga Instructor   

                                             job_summary  summary_len  \
0      Wolfoods operates summer camp kitchens across ...         5863   
1      Wolfoods operates summer camp kitchens across ...         5593   
2      BE PART OF SOMETHING\nREAL\nYOUR ROLE\nAs the ...         4932   
3      BE PART OF SOMETHING\nREAL\nYOUR ROLE\nAs the ...         4942   
4      BE PART OF SOMETHING\nREAL\nYOUR ROLE\nAs the ...         4944   
...                                                  ...          ...   
37195  Position Title:\nYoga Instructor\nLocation:\nV...         2953   
37196  Yoga Teacher for Vinyasa-Based Heated Yoga & S...         3782   
37197  Yoga Teacher for Vinyasa-Based Heated Yoga & S...         2030   
37198  Yoga Teacher for Vinyasa-Based Heated Yoga & S...         2213   
37199  Yoga Teacher for Vinyasa-Based Heated Yoga & S...         2065   

       skill_count  
0               69  
1               69  
2               20  
3               35  
4               27  
...            ...  
37195            8  
37196           15  
37197           12  
37198           23  
37199           22  

[37200 rows x 4 columns]

In [12]:
print("Number of unique job titles:", resp_df['job_title'].nunique())

Number of unique job titles: 1168


The plan is to obtain 1500 records for generating synthetic resumes. Therefore, it would be better to choose the job_summary that has the most information about the job. So, for each job title, I will select the record with max. summary_len. That makes 1168 records. For the rest of the records, I will select some famous job titles like Data Engineer, Data Scientist, etc., that have the highest number of skill_count

## Final Sampling

### Select limited job titles

In [13]:
selected_titles = ['Data Engineer', 'Senior Data Engineer', 'Lead Data Engineer', 'Data Analyst', 'Senior Data Analyst', 'Business Intelligence Analyst', 'Data Scientist', 'Senior Data Scientist', 'Machine Learning Engineer', 'Senior Machine Learning Engineer', 'Software Engineer', 'Senior Software Engineer', 'Software Developer', 'Senior Software Developer', 'Full Stack Developer', 'Senior Full Stack Developer', 'Senior Fullstack Engineer', 'DevOps Engineer', 'Senior DevOps Engineer', 'System Engineer', 'Senior System Engineer', 'Systems Engineer', 'Senior Systems Engineer', 'Senior Lead Engineer - Generative AI Product Engineering (Remote-Eligible)', 'UX Designer', 'Product Designer', 'Senior Product Designer', 'Quality Engineer', 'Senior Quality Engineer', 'Quality Assurance Engineer', 'Research Scientist', 'Senior Research Scientist', 'Applied Researcher I', 'Applied Researcher II', 'Distinguished Applied Researcher', 'Test Engineer', 'Senior Test Engineer', 'Automation Engineer', 'Cloud Engineer', 'Security Engineer', 'Senior Security Engineer', 'Site Reliability Engineer', 'Senior Site Reliability Engineer', 'AI Engineer', 'ML Engineer', 'Senior MLOps Engineer', 'Computer Vision Engineer', 'Senior Computer Vision Engineer', 'Embedded Software Engineer', 'Senior Embedded Software Engineer', 'Technical Architect', 'Solution Architect', 'Solutions Architect', 'Enterprise Architect', 'Technical Lead, Software Engineering', 'Blockchain Engineer', 'Robotics Engineer', 'Senior Robotics Engineer'] + ['Senior Systems Engineer', 'Senior Quality Engineer', 'Software Engineer', 'Senior Data Engineer', 'Senior Software Developer', 'Business Intelligence Analyst', 'Solutions Architect', 'Senior Data Scientist', 'Senior Data Analyst', 'Data Engineer', 'Senior Test Engineer', 'DevOps Engineer', 'Registered Nurse', 'Nurse Practitioner', 'Physical Therapist', 'Medical Technologist', 'Licensed Mental Health Counselor', 'Physician', 'Clinical Laboratory Scientist', 'Financial Analyst', 'Senior Tax Accountant', 'Project Coordinator', 'Business Development Representative', 'Marketing Coordinator', 'Audit Manager', 'Retail Store Manager', 'Sales Associate', 'Customer Service Representative', 'Restaurant Manager', 'Food Service Supervisor', 'Sales Manager', 'Retail Assistant Store Manager', 'Maintenance Technician', 'Quality Control Inspector', 'Production Planner', 'CNC Machinist', 'Construction Supervisor', 'Welder', 'Electrician', 'Operations Manager', 'Facilities Manager', 'HR Manager', 'Supply Chain Manager', 'Project Manager', 'Environmental Health & Safety Manager', 'Plant Controller']

print(len(selected_titles))

104


In [14]:
selected_jd = resp_df[resp_df['job_title'].isin(selected_titles)]
selected_jd

job_link  \
26      https://www.linkedin.com/jobs/view/physical-th...   
45      https://www.linkedin.com/jobs/view/project-coo...   
60      https://www.linkedin.com/jobs/view/sales-manag...   
63      https://www.linkedin.com/jobs/view/food-servic...   
76      https://www.linkedin.com/jobs/view/marketing-c...   
...                                                   ...   
125304  https://www.linkedin.com/jobs/view/production-...   
125308  https://www.linkedin.com/jobs/view/plant-contr...   
125311  https://www.linkedin.com/jobs/view/supply-chai...   
125340  https://www.linkedin.com/jobs/view/supply-chai...   
125375  https://uk.linkedin.com/jobs/view/production-p...   

                  last_processed_time got_summary got_ner is_being_worked  \
26      2024-01-21 08:08:52.028232+00           t       t               f   
45      2024-01-21 08:08:28.123873+00           t       t               f   
60      2024-01-21 07:39:57.333128+00           t       t               f   
63      2024-01-21 08:08:34.452912+00           t       t               f   
76      2024-01-21 08:20:46.482342+00           t       t               f   
...                               ...         ...     ...             ...   
125304  2024-01-21 07:39:32.812628+00           t       t               f   
125308  2024-01-21 00:38:29.061053+00           t       t               f   
125311  2024-01-21 07:11:54.013677+00           t       t               f   
125340  2024-01-21 08:20:26.354909+00           t       t               f   
125375  2024-01-20 04:01:08.913697+00           t       t               f   

                      job_title  \
26           Physical Therapist   
45          Project Coordinator   
60                Sales Manager   
63      Food Service Supervisor   
76        Marketing Coordinator   
...                         ...   
125304       Production Planner   
125308         Plant Controller   
125311     Supply Chain Manager   
125340     Supply Chain Manager   
125375       Production Planner   

                                              company  \
26                                   Hydro Recruiting   
45      Facilities Maintenance Management, LLC. (FMM)   
60                         Crestline Hotels & Resorts   
63                                 Global Partners LP   
76                                      Oldcastle APG   
...                                               ...   
125304                      Terra Sonic International   
125308                                          Jobot   
125311                                    Cypress HCM   
125340                   Staffing Resource Group, Inc   
125375                Eat Real + PROPER | WARP Snacks   

                             job_location  first_seen       search_city  \
26                          Jonesboro, AR  2024-01-12          Arkansas   
45                     Denham Springs, LA  2024-01-13         Louisiana   
60                              Tampa, FL  2024-01-13          Seminole   
63                         West Haven, CT  2024-01-14             Derby   
76                         Minnetonka, MN  2024-01-14  South Saint Paul   
...                                   ...         ...               ...   
125304                       Marietta, OH  2024-01-16              Ohio   
125308                       Lawrence, KS  2024-01-16            Kansas   
125311                         Totowa, NJ  2024-01-16         Montclair   
125340                      Billerica, MA  2024-01-14           Ipswich   
125375  Nuneaton, England, United Kingdom  2024-01-14          Hastings   

        search_country               search_position   job_level job_type  \
26       United States                 Acupuncturist  Mid senior   Onsite   
45       United States           Program Coordinator  Mid senior   Onsite   
60       United States                 Range Manager  Mid senior   Onsite   
63       United States                       Starter  Mid senior   Onsite   

In [15]:
# Group by job_title and company, then find index of row with max summary_len in each group
idx = selected_jd.groupby(['job_title', 'company'])['summary_len'].idxmax()

# Use these indices to filter the dataframe
filtered_df = selected_jd.loc[idx]

# Reset index if needed
filtered_df = filtered_df.reset_index(drop=True)
filtered_df

job_link  \
0     https://www.linkedin.com/jobs/view/applied-res...   
1     https://www.linkedin.com/jobs/view/applied-res...   
2     https://uk.linkedin.com/jobs/view/audit-manage...   
3     https://uk.linkedin.com/jobs/view/audit-manage...   
4     https://uk.linkedin.com/jobs/view/audit-manage...   
...                                                 ...   
2248  https://www.linkedin.com/jobs/view/welder-at-n...   
2249  https://www.linkedin.com/jobs/view/welder-at-p...   
2250  https://www.linkedin.com/jobs/view/welder-at-s...   
2251  https://www.linkedin.com/jobs/view/welder-at-t...   
2252  https://www.linkedin.com/jobs/view/welder-at-w...   

                last_processed_time got_summary got_ner is_being_worked  \
0     2024-01-20 14:49:25.961976+00           t       t               f   
1     2024-01-21 08:39:43.241279+00           t       t               f   
2     2024-01-19 22:14:32.073908+00           t       t               f   
3     2024-01-19 09:45:09.215838+00           t       t               f   
4     2024-01-19 09:45:09.215838+00           t       t               f   
...                             ...         ...     ...             ...   
2248  2024-01-19 09:45:09.215838+00           t       t               f   
2249  2024-01-19 09:45:09.215838+00           t       t               f   
2250  2024-01-19 16:32:53.648536+00           t       t               f   
2251  2024-01-19 09:45:09.215838+00           t       t               f   
2252  2024-01-19 09:45:09.215838+00           t       t               f   

                  job_title                              company  \
0      Applied Researcher I                    Jobs for Humanity   
1     Applied Researcher II                    Jobs for Humanity   
2             Audit Manager                                  8-i   
3             Audit Manager                         ACCA Careers   
4             Audit Manager                              AESC UK   
...                     ...                                  ...   
2248                 Welder              NorthPoint Search Group   
2249                 Welder  Phoenix Fabricators & Erectors, LLC   
2250                 Welder                 Satellite Industries   
2251                 Welder         Textron Specialized Vehicles   
2252                 Welder                 Waltonen Engineering   

                             job_location  first_seen      search_city  \
0                             Concord, NH  2024-01-15       Manchester   
1                        Poughkeepsie, NY  2024-01-15      New Windsor   
2     Canterbury, England, United Kingdom  2024-01-14  Southend-on-Sea   
3      Stevenage, England, United Kingdom  2024-01-13            Luton   
4     Sunderland, England, United Kingdom  2024-01-12         Tyneside   
...                                   ...         ...              ...   
2248                            Byron, GA  2024-01-13            Macon   
2249                             Avon, IN  2024-01-13       Plainfield   
2250                        Dahlonega, GA  2024-01-14        Dahlonega   
2251                     Cartersville, GA  2024-01-13          Georgia   
2252                             Troy, MI  2024-01-13         Oak Park   

      search_country               search_position   job_level job_type  \
0      United States             Director Research  Mid senior   Onsite   
1      United States             Director Research  Mid senior   Onsite   
2     United Kingdom                       Auditor  Mid senior   Onsite   
3     United Kingdom                Director Stage  Mid senior   Onsite   
4     United Kingdom                       Auditor  Mid senior   Onsite   
...              ...                           ...         ...      ...   
2248   United States  Electromechanical Technician  Mid senior   Onsite   
2249   United States  Electromechanical Technician   Associate   Onsite   
2250   United States  Electromechanical Technician  Mid 

In [16]:
# Step 1: Count the occurrences of each job_title
job_title_counts = filtered_df['job_title'].value_counts()

# Step 2: Filter to only include job titles with more than 15 records
valid_job_titles = job_title_counts[job_title_counts >= 15].index.tolist()

# Step 3: Filter the original dataframe to only include these job titles
result_df = filtered_df[filtered_df['job_title'].isin(valid_job_titles)].copy()
result_df

job_link  \
2     https://uk.linkedin.com/jobs/view/audit-manage...   
3     https://uk.linkedin.com/jobs/view/audit-manage...   
4     https://uk.linkedin.com/jobs/view/audit-manage...   
5     https://www.linkedin.com/jobs/view/audit-manag...   
6     https://www.linkedin.com/jobs/view/audit-manag...   
...                                                 ...   
2222  https://www.linkedin.com/jobs/view/test-engine...   
2223  https://www.linkedin.com/jobs/view/test-engine...   
2224  https://www.linkedin.com/jobs/view/test-engine...   
2225  https://www.linkedin.com/jobs/view/test-engine...   
2226  https://www.linkedin.com/jobs/view/test-engine...   

                last_processed_time got_summary got_ner is_being_worked  \
2     2024-01-19 22:14:32.073908+00           t       t               f   
3     2024-01-19 09:45:09.215838+00           t       t               f   
4     2024-01-19 09:45:09.215838+00           t       t               f   
5     2024-01-19 09:45:09.215838+00           t       t               f   
6     2024-01-20 10:01:02.122191+00           t       t               f   
...                             ...         ...     ...             ...   
2222  2024-01-19 09:45:09.215838+00           t       t               f   
2223  2024-01-19 09:45:09.215838+00           t       t               f   
2224  2024-01-19 09:45:09.215838+00           t       t               f   
2225  2024-01-19 09:45:09.215838+00           t       t               f   
2226  2024-01-19 09:45:09.215838+00           t       t               f   

          job_title                             company  \
2     Audit Manager                                 8-i   
3     Audit Manager                        ACCA Careers   
4     Audit Manager                             AESC UK   
5     Audit Manager                   Alera Group, Inc.   
6     Audit Manager                   Ascendo Resources   
...             ...                                 ...   
2222  Test Engineer  Teledyne Microwave Solutions (TMS)   
2223  Test Engineer              Trojan Battery Company   
2224  Test Engineer                   US Tech Solutions   
2225  Test Engineer                   Viking Range, LLC   
2226  Test Engineer                   nFocus Solutions®   

                                      job_location  first_seen  \
2              Canterbury, England, United Kingdom  2024-01-14   
3               Stevenage, England, United Kingdom  2024-01-13   
4              Sunderland, England, United Kingdom  2024-01-12   
5                                    Manasquan, NJ  2024-01-12   
6     Springfield, Massachusetts Metropolitan Area  2024-01-14   
...                                            ...         ...   
2222                             Mountain View, CA  2024-01-13   
2223                                   Horsham, PA  2024-01-13   
2224                                     Tempe, AZ  2024-01-13   
2225                                 Greenwood, MS  2024-01-13   
2226                                   Orlando, FL  2024-01-13   

          search_city  search_country                    search_position  \
2     Southend-on-Sea  United Kingdom                            Auditor   
3               Luton  United Kingdom                     Director Stage   
4            Tyneside  United Kingdom                            Auditor   
5         Asbury Park   United States                            Auditor   
6         Springfield   United States                            Auditor   
...               ...             ...                                ...   
2222      Santa Clara   United States          Electronics-Test Engineer   
2223       Norristown   United States  Computer Systems Hardware Analyst   
2224         Glendale   United States                  Computer Operator   
2225      Mississippi   United States                Commercial Engineer   
2226         Avondale   United States                  Computer Operator   

       job_level job_t

In [17]:
# Step 4: Select top 16 records per job title based on summary_len
result_df = (result_df
             .groupby('job_title')
             .apply(lambda x: x.nlargest(10, 'summary_len'))
             .reset_index(drop=True)).copy()

result_df

/var/folders/t5/tflj5hjs197_vv37fw2sdhdr0000gn/T/ipykernel_57662/2125768498.py:4: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda x: x.nlargest(10, 'summary_len'))


job_link  \
0    https://uk.linkedin.com/jobs/view/audit-manage...   
1    https://www.linkedin.com/jobs/view/audit-manag...   
2    https://www.linkedin.com/jobs/view/audit-manag...   
3    https://www.linkedin.com/jobs/view/audit-manag...   
4    https://uk.linkedin.com/jobs/view/audit-manage...   
..                                                 ...   
495  https://www.linkedin.com/jobs/view/test-engine...   
496  https://www.linkedin.com/jobs/view/test-engine...   
497  https://www.linkedin.com/jobs/view/test-engine...   
498  https://www.linkedin.com/jobs/view/test-engine...   
499  https://www.linkedin.com/jobs/view/test-engine...   

               last_processed_time got_summary got_ner is_being_worked  \
0    2024-01-19 09:45:09.215838+00           t       t               f   
1    2024-01-19 15:33:11.588617+00           t       t               f   
2    2024-01-19 09:45:09.215838+00           t       t               f   
3    2024-01-19 09:45:09.215838+00           t       t               f   
4    2024-01-20 06:56:42.454545+00           t       t               f   
..                             ...         ...     ...             ...   
495   2024-01-19 20:05:43.91119+00           t       t               f   
496  2024-01-19 09:45:09.215838+00           t       t               f   
497  2024-01-19 09:45:09.215838+00           t       t               f   
498  2024-01-19 09:45:09.215838+00           t       t               f   
499  2024-01-19 09:45:09.215838+00           t       t               f   

         job_title                               company  \
0    Audit Manager                          ACCA Careers   
1    Audit Manager                        Baker Tilly US   
2    Audit Manager                                 KORE1   
3    Audit Manager                                 Jobot   
4    Audit Manager  Bishop Fleming Chartered Accountants   
..             ...                                   ...   
495  Test Engineer                           Stoke Space   
496  Test Engineer                               Akkodis   
497  Test Engineer                     nFocus Solutions®   
498  Test Engineer                    Micross Components   
499  Test Engineer                   SEGULA Technologies   

                           job_location  first_seen     search_city  \
0    Stevenage, England, United Kingdom  2024-01-13           Luton   
1                       Minneapolis, MN  2024-01-15           Orono   
2                        Cincinnati, OH  2024-01-12      Cincinnati   
3                          Cheyenne, WY  2024-01-12         Wyoming   
4      Bristol, England, United Kingdom  2024-01-14         Bristol   
..                                  ...         ...             ...   
495                      Moses Lake, WA  2024-01-14         Ephrata   
496                           Tempe, AZ  2024-01-12          Canyon   
497                         Orlando, FL  2024-01-13        Avondale   
498                    Los Alamitos, CA  2024-01-12        Pasadena   
499                      Greensboro, NC  2024-01-12  North Carolina   

     search_country                 search_position   job_level job_type  \
0    United Kingdom                  Director Stage  Mid senior   Onsite   
1     United States                           Baker  Mid senior   Onsite   
2     United States                         Auditor  Mid senior   Onsite   
3     United States                         Auditor  Mid senior   Onsite   
4    United Kingdom                            Tier  Mid senior   Onsite   
..              ...                             ...         ...      ...   
495   United States  Agricultural-Research Engineer  Mid senior   Onsite   
496   United States  Agricultural-Research Engineer   Associate   Onsite   
497   United States               Computer Operator  Mid senior   Onsite   
498   United States  Agricultural-Research Engineer   Associate   Onsite   
499   United States  Agricultural-Research Enginee

In [18]:
result_df[['skill_count']].describe()

skill_count
count    500.00000
mean      30.60400
std       13.97203
min        9.00000
25%       21.00000
50%       28.00000
75%       37.00000
max      130.00000

In [19]:
result_df.head(250).to_csv(f"{path}/data/sampled_linkedin_jd_skills/sampled_data_v2/sampled_jd_resume_set1.csv", index=False)
result_df.tail(250).to_csv(f"{path}/data/sampled_linkedin_jd_skills/sampled_data_v2/sampled_jd_resume_set2.csv", index=False)